src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpFormattingPassBase.cs

```diff
@@ -108,7 +108,8 @@ protected async Task<List<TextChange>> AdjustIndentationAsync(FormattingContext
             var significantLocationIndentation = await CSharpFormatter.GetCSharpIndentationAsync(context, significantLocations, cancellationToken);
 
             // Build source mapping indentation scopes.
-            var sourceMappingIndentations = new SortedDictionary<int, int>();
+            var sourceMappingIndentations = new SortedDictionary<int, (int cSharpDesiredIndentation, RazorDirectiveSyntax? containingDirective)>();
```

Comment:
We only use the containing directive to get its end position, and when building this dictionary, we are already aware of positions anyway, so I wonder if its possible to simply artificially add a new entry to the dictionary, when we get to the end of the directive. ie, instead of the dictionary being something like:

```
[1, (12, null)] // naturally occuring source mapping
[10, (16, directive)] // naturally occuring source mapping, for a directive that has an end location in position 18
[18, (12, null)] // artificially inserted source mapping indentation point, reverting indentation back to what it was before the directive
[24, (12, null)] // next naturally occuring source mapping
```

Obviously after this is done, we wouldn't need the tuple at all, and in fact the rest of the code in this method could be left as is. All of which is to say, can we build a smarter map of indentation scopes, and process them simply, rather than trying to be smart about how we process a map that doesn't have enough information in it.